src/DataLayer/dataSources/DataSource.ts

```diff
@@ -1,25 +1,95 @@
 // (C) 2007-2018 GoodData Corporation
-import { AFM } from '@gooddata/typings';
+import { AFM, Execution } from '@gooddata/typings';
 import * as stringify from 'json-stable-stringify';
 import { IDataSource } from '../interfaces/DataSource';
 
 export {
     IDataSource
 };
 
+export const SERVER_SIDE_COLUMN_LIMIT = 1000;
+export const SERVER_SIDE_ROW_LIMIT = 1000;
+
+const ROW_DIMENSION_INDEX = 0;
+const COLUMN_DIMENSION_INDEX = 1;
+
 export type ExecFactory<T> = (resultSpec: AFM.IResultSpec) => Promise<T>;
 
 export class DataSource<T> implements IDataSource<T> {
+    private executionPromises: {
+        [key: string]: Promise<Execution.IExecutionResponse>
+    };
+
     constructor(
         private execFactory: ExecFactory<T>,
         private afm: AFM.IAfm,
-        private fingerprint?: string
-    ) {}
+        private fingerprint?: string,
+        private responseFactory?: (resultSpec: AFM.IResultSpec) => Promise<Execution.IExecutionResponse>,
```

Comment:
This looks to me like something wrong design

I would expect here new `PageableDataSource` inherited from `DataSource`

what is the reason to not doing it this way?

Reply:
No changes right now as discussed with @tomasreichmann